core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -173,16 +170,39 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * then the action will fail with some error.
    */
   final def raceBoth[E1 >: E, B](that: IO[E1, B]): IO[E1, Either[A, B]] =
-    raceWith(that)((a, fiber) => fiber.interrupt.const(Left(a)), (b, fiber) => fiber.interrupt.const(Right(b)))
+    raceWith(that)(_.join.map(Left(_)) <* _.interrupt, _.join.map(Right(_)) <* _.interrupt)
 
   /**
    * Races this action with the specified action, invoking the
    * specified finisher as soon as one value or the other has been computed.
    */
-  final def raceWith[E1 >: E, A1 >: A, B, C](
+  final def raceWith[E1, E2, B, C](
     that: IO[E1, B]
-  )(finishLeft: (A1, Fiber[E1, B]) => IO[E1, C], finishRight: (B, Fiber[E1, A1]) => IO[E1, C]): IO[E1, C] =
-    new IO.Race[E1, A1, B, C](self, that, finishLeft, finishRight)
+  )(
+    leftWins: (Fiber[E, A], Fiber[E1, B]) => IO[E2, C],
+    rightWins: (Fiber[E1, B], Fiber[E, A]) => IO[E2, C]
+  ): IO[E2, C] = {
+    def arbiter[E0, E1, A, B](
+      f: (Fiber[E0, A], Fiber[E1, B]) => IO[E2, C],
+      winner: Fiber[E0, A],
+      loser: Fiber[E1, B],
+      race: Ref[IO.Race],
+      done: Promise[E2, C]
+    )(res: ExitResult[E0, A]): IO[Nothing, _] =
+      race
+        .modify(r => r.won(res.succeeded) -> r.next(res.succeeded))
+        .flatMap(IO.when(_)(f(winner, loser).run.flatMap(done.done(_)).void))
```

Comment:
This makes me want to add:

```scala
trait IO[+E, +A] {
  final def to(p: Promise[E, A]): IO[Nothing, Unit] = 
    self.run.flatMap(p.done(_)).void
}
```

And of course it makes me want to optimize `run`. 😄 

This looks great to me! It's very tricky to get this perfectly right.